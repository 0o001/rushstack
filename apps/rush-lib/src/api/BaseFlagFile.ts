// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { FileSystem, JsonFile, JsonObject } from '@rushstack/node-core-library';

/**
 * A helper class for managing last-install flags, which are persistent and
 * indicate that something installed in the folder was successfully completed.
 * It also compares state, so that if something like the Node.js version has changed,
 * it can invalidate the last install.
 * @internal
 */
export abstract class BaseFlagFile {
  private _path: string;
  private _state: JsonObject;

  /**
   * Creates a new BaseFlagFile
   *
   * @param flagPath - the file containing the flag information
   * @param state - optional, the state that should be managed or compared
   */
  protected constructor(flagPath: string, state: JsonObject = {}) {
    this._path = flagPath;
    this._state = state;
  }

  /**
   * Returns the full path to the flag file
   */
  public get path(): string {
    return this._path;
  }

  /**
   * Returns the current flag state
   */
  protected get state(): JsonObject {
    return this._state;
  }

  /**
   * Returns true if the file exists and the contents match the current state.
   */
  public abstract isValid(): boolean;

  /**
   * Writes the flag file to disk with the current state
   */
  public create(): void {
    JsonFile.save(this._state, this._path, {
      ensureFolderExists: true
    });
  }

  /**
   * Removes the flag file
   */
  public clear(): void {
    FileSystem.deleteFile(this._path);
  }

  /**
   * Load and return the flag from file, or undefined if the flag could not be loaded.
   */
  protected loadFromFile(): JsonObject | undefined {
    try {
      return JsonFile.load(this.path);
    } catch (err) {
      return;
    }
  }
}
