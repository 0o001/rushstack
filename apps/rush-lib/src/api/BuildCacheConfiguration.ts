// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';
import { JsonFile, JsonSchema, FileSystem } from '@rushstack/node-core-library';

import { BuildCacheProviderBase } from '../logic/buildCache/BuildCacheProviderBase';
import {
  AzureEnvironment,
  AzureEnvironmentNames,
  AzureStorageBuildCacheProvider
} from '../logic/buildCache/AzureStorageBuildCacheProvider';
import { RushConfiguration } from './RushConfiguration';
import { FileSystemBuildCacheProvider } from '../logic/buildCache/FileSystemBuildCacheProvider';
import { BuildCacheProviderCredentialCache } from '../logic/buildCache/BuildCacheProviderCredentialCache';
import { RushGlobalFolder } from './RushGlobalFolder';

/**
 * Describes the file structure for the "common/config/rush/build-cache.json" config file.
 */
interface IBuildCacheJson {
  cacheProvider: 'azure-blob-storage' | 'filesystem';

  /**
   * A list of folder names under each project root that should be cached.
   * These folders should not be tracked by git.
   */
  projectOutputFolderNames: string[];
}

interface IAzureBlobStorageBuildCacheJson extends IBuildCacheJson {
  cacheProvider: 'azure-blob-storage';

  azureBlobStorageConfiguration: IAzureStorageConfigurationJson;
}

interface IAzureStorageConfigurationJson {
  /**
   * The name of the the Azure storage account to use for build cache.
   */
  storageAccountName: string;

  /**
   * The name of the container in the Azure storage account to use for build cache.
   */
  storageContainerName: string;

  /**
   * The Azure environment the storage account exists in. Defaults to AzureCloud.
   */
  azureEnvironment?: AzureEnvironmentNames;

  /**
   * An optional prefix for cache item blob names.
   */
  blobPrefix?: string;

  /**
   * If set to true, allow writing to the cache. Defaults to false.
   */
  isCacheWriteAllowed?: boolean;
}

interface IFileSystemBuildCacheJson extends IBuildCacheJson {
  cacheProvider: 'filesystem';
}

/**
 * Use this class to load and save the "common/config/rush/build-cache.json" config file.
 * This file provides configuration options for cached project build output.
 * @public
 */
export class BuildCacheConfiguration {
  private static _jsonSchema: JsonSchema = JsonSchema.fromFile(
    path.join(__dirname, '..', 'schemas', 'build-cache.schema.json')
  );

  public readonly projectOutputFolderNames: string[];

  public readonly cacheProvider: BuildCacheProviderBase;

  private constructor(
    buildCacheJson: IBuildCacheJson,
    rushConfiguration: RushConfiguration,
    credentialCache: BuildCacheProviderCredentialCache
  ) {
    this.projectOutputFolderNames = buildCacheJson.projectOutputFolderNames;

    switch (buildCacheJson.cacheProvider) {
      case 'filesystem': {
        this.cacheProvider = new FileSystemBuildCacheProvider({
          credentialCache,
          rushConfiguration
        });
        break;
      }

      case 'azure-blob-storage': {
        const azureStorageBuildCacheJson: IAzureBlobStorageBuildCacheJson = buildCacheJson as IAzureBlobStorageBuildCacheJson;
        const azureStorageConfigurationJson: IAzureStorageConfigurationJson =
          azureStorageBuildCacheJson.azureBlobStorageConfiguration;
        const azureEnvironment: AzureEnvironment | undefined = azureStorageConfigurationJson.azureEnvironment
          ? AzureStorageBuildCacheProvider.parseAzureEnvironmentName(
              azureStorageConfigurationJson.azureEnvironment
            )
          : undefined;
        this.cacheProvider = new AzureStorageBuildCacheProvider({
          credentialCache,
          storageAccountName: azureStorageConfigurationJson.storageAccountName,
          storageContainerName: azureStorageConfigurationJson.storageContainerName,
          azureEnvironment: azureEnvironment,
          blobPrefix: azureStorageConfigurationJson.blobPrefix,
          isCacheWriteAllowed: !!azureStorageConfigurationJson.isCacheWriteAllowed
        });
        break;
      }

      default: {
        throw new Error(`Unexpected cache provider: ${buildCacheJson.cacheProvider}`);
      }
    }
  }

  /**
   * Loads the build-cache.json data from the specified file path.
   * If the file has not been created yet, then undefined is returned.
   */
  public static async loadFromFileAsync(
    jsonFilename: string,
    rushConfiguration: RushConfiguration,
    rushGlobalFolder: RushGlobalFolder
  ): Promise<BuildCacheConfiguration | undefined> {
    if (FileSystem.exists(jsonFilename)) {
      const buildCacheJsonPromise: Promise<IBuildCacheJson> = JsonFile.loadAndValidateAsync(
        jsonFilename,
        BuildCacheConfiguration._jsonSchema
      );
      const credentialCachePromise: Promise<BuildCacheProviderCredentialCache> = BuildCacheProviderCredentialCache.initializeAsync(
        rushGlobalFolder
      );
      const [buildCacheJson, credentialCache] = await Promise.all([
        buildCacheJsonPromise,
        credentialCachePromise
      ]);

      return new BuildCacheConfiguration(buildCacheJson, rushConfiguration, credentialCache);
    } else {
      return undefined;
    }
  }
}
