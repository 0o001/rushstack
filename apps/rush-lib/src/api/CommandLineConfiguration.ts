// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';
import * as lodash from 'lodash';

import {
  JsonFile,
  JsonSchema,
  FileSystem
} from '@microsoft/node-core-library';

import { RushConstants } from '../logic/RushConstants';

import {
  CommandJson,
  ICommandLineJson,
  ParameterJson
} from './CommandLineJson';

/**
 * Custom Commands and Options for the Rush Command Line
 */
export class CommandLineConfiguration {
  private static _jsonSchema: JsonSchema = JsonSchema.fromFile(
    path.join(__dirname, '../schemas/command-line.schema.json'));

  public readonly commands: CommandJson[] = [];
  public readonly parameters: ParameterJson[] = [];

  /**
   * Loads the configuration from the specified file.  If the file does not exist,
   * then an empty default instance is returned.  If the file contains errors, then
   * an exception is thrown.
   */
  public static loadFromFileOrDefault(jsonFilename: string): CommandLineConfiguration {
    let commandLineJson: ICommandLineJson | undefined = undefined;
    if (FileSystem.exists(jsonFilename)) {
      commandLineJson = JsonFile.loadAndValidate(jsonFilename, CommandLineConfiguration._jsonSchema);
    }

    return new CommandLineConfiguration(commandLineJson);
  }

  /**
   * Loads the configuration from the specified file and applies any ommited default build
   * settings.  If the file does not exist, then an empty default instance is returned.
   * If the file contains errors, then an exception is thrown.
   */
  public static loadFromFileAndDefault(jsonFilename: string): CommandLineConfiguration {
    let commandLineJson: ICommandLineJson | undefined = undefined;
    if (FileSystem.exists(jsonFilename)) {
      commandLineJson = JsonFile.load(jsonFilename);

      // TODO: Investigate
      const defaultBuildSettings: CommandJson[] =
        [{
          commandKind: 'bulk',
          name: 'build',
          summary: '(EXPERIMENTAL) Build all projects that haven\'t been built, or have changed since they were last'
          + ' built.',
          enableParallelism: true,
          ignoreMissingScript: false,
          ignoreDependencyOrder: false,
          allowWarningsInSuccessfulBuild: false,
          safeForSimultaneousRushProcesses: false
        },
        {
          commandKind: 'bulk',
          name: 'rebuild',
          summary: 'Clean and rebuild the entire set of projects',
          enableParallelism: true,
          ignoreMissingScript: false,
          ignoreDependencyOrder: false,
          allowWarningsInSuccessfulBuild: false,
          safeForSimultaneousRushProcesses: false
        }
      ];

      const applyDefaults: boolean = commandLineJson ? lodash.some(commandLineJson.commands, (command) => {
        return command.commandKind === 'bulk' && (command.name === 'build' || command.name === 'rebuild');
      }) : false;

      let mergedBuildSettings: CommandJson[] = [];
      // merge commands specified in command-line.json and default (re)build settings
      // Ensure both build commands are included and preserve any other commands specified
      if (commandLineJson) {
        if (applyDefaults) {
          mergedBuildSettings = lodash({}) // Start with an empty object
          .merge(
            lodash(defaultBuildSettings).groupBy('name').value(),
            lodash(commandLineJson.commands).groupBy('name').value()
            )
            .values()
            .flatten()
            .value();
          commandLineJson.commands = mergedBuildSettings;
        }
        CommandLineConfiguration._jsonSchema.validateObject(commandLineJson, jsonFilename);
      }
    }

    return new CommandLineConfiguration(commandLineJson);

  }

  /**
   * Use CommandLineConfiguration.loadFromFile()
   */
  private constructor(commandLineJson: ICommandLineJson | undefined) {
    if (commandLineJson) {
      if (commandLineJson.commands) {
        for (const command of commandLineJson.commands) {
          this.commands.push(command);
        }
      }

      if (commandLineJson.parameters) {
        for (const parameter of commandLineJson.parameters) {
          this.parameters.push(parameter);

          // Do some basic validation
          switch (parameter.parameterKind) {
            case 'choice':
              const alternativeNames: string[] = parameter.alternatives.map(x => x.name);

              if (parameter.defaultValue && alternativeNames.indexOf(parameter.defaultValue) < 0) {
                throw new Error(`In ${RushConstants.commandLineFilename}, the parameter "${parameter.longName}",`
                  + ` specifies a default value "${parameter.defaultValue}"`
                  + ` which is not one of the defined alternatives: "${alternativeNames.toString()}"`);
              }
              break;
          }
        }
      }

    }
  }
}
