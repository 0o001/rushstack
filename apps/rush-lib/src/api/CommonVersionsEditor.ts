// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { IDependencyFileEditor, PackageJsonDependency, DependencyType } from './PackageJsonEditor';
import { CommonVersionsConfiguration } from './CommonVersionsConfiguration';
import { RushConstants } from '../logic/RushConstants';

export class CommonVersionsEditor implements IDependencyFileEditor {
  private _commonVersionsConfiguration: CommonVersionsConfiguration;

  public constructor(commonVersionsConfiguration: CommonVersionsConfiguration) {
    this._commonVersionsConfiguration = commonVersionsConfiguration;
  }

  public get filePath(): string {
    return this._commonVersionsConfiguration.filePath;
  }

  public get allDependencies(): ReadonlyArray<PackageJsonDependency> {
    const dependencies: PackageJsonDependency[] = [];

    this._commonVersionsConfiguration.getAllPreferredVersions().forEach((version, dependencyName) => {
      dependencies.push(this._getPackageJsonDependency(dependencyName, version));
    });

    return dependencies;
  }

  public tryGetDependency(packageName: string): PackageJsonDependency | undefined {
    const version: string | undefined = this._commonVersionsConfiguration.getAllPreferredVersions().get(packageName);
    if (!version) {
      return undefined;
    } else {
      return this._getPackageJsonDependency(packageName, version);
    }
  }

  public tryGetDevDependency(packageName: string): PackageJsonDependency | undefined {
    return undefined; // common-versions.json doesn't have a distinction between dev and non-dev dependencies
  }

  public addOrUpdateDependency(packageName: string, newVersion: string, dependencyType: DependencyType): void {
    if (dependencyType !== DependencyType.Regular) {
      throw new Error(`${RushConstants.commonVersionsFilename} only accepts "${DependencyType.Regular}" dependencies`);
    }

    if (this._commonVersionsConfiguration.xstitchPreferredVersions.has(packageName)) {
      this._commonVersionsConfiguration.xstitchPreferredVersions.set(packageName, newVersion);
    } else {
      this._commonVersionsConfiguration.preferredVersions.set(packageName, newVersion);
    }
  }

  public saveIfModified(): boolean {
    return this._commonVersionsConfiguration.save();
  }

  private _getPackageJsonDependency(dependencyName: string, version: string): PackageJsonDependency {
    return new PackageJsonDependency(
      dependencyName,
      version,
      DependencyType.Regular,
      () => this.addOrUpdateDependency(dependencyName, version, DependencyType.Regular)
    );
  }
}
