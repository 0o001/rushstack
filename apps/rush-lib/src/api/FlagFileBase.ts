// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { FileSystem, JsonFile } from '@rushstack/node-core-library';

/**
 * A helper class for managing persistent flag files. It also compares state, so that if
 * some tracked state has changed, it can invalidate the flag file.
 * @internal
 */
export abstract class FlagFileBase<TState> {
  private _path: string;
  private _state: TState;

  /**
   * Creates a new FlagFileBase
   *
   * @param flagPath - the file containing the flag information
   * @param state - optional, the state that should be managed or compared
   */
  protected constructor(flagPath: string, state: TState) {
    this._path = flagPath;
    this._state = state;
  }

  /**
   * Returns the full path to the flag file
   */
  public get path(): string {
    return this._path;
  }

  /**
   * Returns the current flag state
   */
  protected get state(): TState {
    return this._state;
  }

  /**
   * Returns true if the file exists and the contents match the current state.
   */
  public abstract isValid(): boolean;

  /**
   * Writes the flag file to disk with the current state
   */
  public create(): void {
    JsonFile.save(this._state, this._path, {
      ensureFolderExists: true
    });
  }

  /**
   * Removes the flag file
   */
  public clear(): void {
    FileSystem.deleteFile(this._path);
  }

  /**
   * Load and return the flag from file, or undefined if the flag could not be loaded.
   */
  protected loadFromFile(): TState | undefined {
    try {
      return JsonFile.load(this.path);
    } catch (err) {
      return;
    }
  }
}
