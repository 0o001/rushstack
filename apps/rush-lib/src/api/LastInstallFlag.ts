import * as path from 'path';
import * as _ from 'lodash';
import { FileSystem, JsonFile, JsonObject } from '@microsoft/node-core-library';

export const LAST_INSTALL_FLAG_FILE_NAME: string = 'last-install.flag';

interface IPropertyErrorDeps {
      /**
       * Original content property must match this value in order to check for
       * inequality
       */
      [prop: string]: string;
}
interface IPropertyErrors {
  /**
   * The properties to check and force an abort on inequality
   */
  [key: string]: {
    /**
     * Properties to check in the original content before checking for inequality.
     * All deps must match.  If no deps are defined, then inequality will be checked
     * unconditionally
     */
    deps?: IPropertyErrorDeps;
    /**
     * In case an inequality is found, and all deps match, then throw this error
     */
    error: string;
  }
}

const PROPERTY_ERRORS: IPropertyErrors = {
  storePath: {
    deps: {
      packageManager: "pnpm"
    },
    error: "Current PNPM store path does not match the last one used.  This may cause inconsistency in your builds.\n\n" +
      "If you wish to install with the currently configured path, please run \"rush update --purge\"\n\n" +
      "Old: <%= oldValue %>\n" +
      "New: <%= newValue %>"
  }
}

/**
 * A helper class for managing last-install flags, which are persistent and
 * indicate that something installed in the folder was sucessfully completed.
 * It also compares state, so that if something like the Node.js version has changed,
 * it can invalidate the last install.
 * @internal
 */
export class LastInstallFlag {
  private _path: string;
  private _state: JsonObject;

  /**
   * Creates a new LastInstall flag
   * @param folderPath - the folder that this flag is managing
   * @param state - optional, the state that should be managed or compared
  */
  public constructor(folderPath: string, state: JsonObject = {}) {
    this._path = path.join(folderPath, LAST_INSTALL_FLAG_FILE_NAME);
    this._state = state;
  }

  /**
   * Returns true if the file exists and the contents match the current state
   * @param abortOnInvalid - If the current state is not equal to the previous
   * state, and an error is defined for the condition that causes inequality, then abort with
   * the defined error message
   */
  public isValid(abortOnInvalid: boolean = false): boolean {
    if (!FileSystem.exists(this._path)) {
      return false;
    }
    let contents: JsonObject;
    try {
      contents = JsonFile.load(this._path);
    } catch (err) {
      return false;
    } 
    if (!_.isEqual(contents, this._state)) {
      if (abortOnInvalid) {
        for (const errorKey of Object.keys(PROPERTY_ERRORS)) {
          const errorDeps: IPropertyErrorDeps = PROPERTY_ERRORS[errorKey].deps || {};
          const depKeys: string[] = Object.keys(errorDeps);
          const depsMatch: boolean = depKeys.length === 0 || depKeys.every((depProp) => errorDeps[depProp] === contents[depProp])
          if (!_.isEqual(contents[errorKey], this._state[errorKey]) && depsMatch) {
            throw _.template(PROPERTY_ERRORS[errorKey].error)({
              oldValue: contents[errorKey],
              newValue: this._state[errorKey]
            });
          }
        }
      }
      return false;
    }
    return true;
  }

  /**
   * Writes the flag file to disk with the current state
   */
  public create(): void {
    JsonFile.save(this._state, this._path, {
      ensureFolderExists: true
    });
  }

  /**
   * Removes the flag file
   */
  public clear(): void {
    FileSystem.deleteFile(this._path);
  }

  /**
   * Returns the full path to the flag file
   */
  public get path(): string {
    return this._path;
  }
}