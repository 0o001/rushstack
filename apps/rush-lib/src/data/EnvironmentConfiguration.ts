// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

const ENV_VALUE_PREFIX: string = 'rush_';

export enum EnvironmentValue {
  TempDirectoryOverride = 'rush_tempDir'
}

const NORMALIZED_ENV_VALUE_NAMES: { [lowercaseValue: string]: EnvironmentValue } = {};
for (const envValue in EnvironmentValue) {
  if (EnvironmentValue.hasOwnProperty(envValue)) {
    const enumValue: string = EnvironmentValue[envValue];
    NORMALIZED_ENV_VALUE_NAMES[enumValue.toLowerCase()] = enumValue as EnvironmentValue;
  }
}

export class EnvironmentConfiguration {
  private static _rushValues: Map<EnvironmentValue, string | undefined> =
    new Map<EnvironmentValue, string| undefined>();

  public static initialize(): void {
    EnvironmentConfiguration._rushValues.clear();

    const unknownEnvVariables: string[] = [];

    for (const envVar in process.env) {
      const normalizedEnvVar: string = envVar.toLowerCase(); // tslint:disable-line:forin
      if (process.env.hasOwnProperty(envVar) && normalizedEnvVar.indexOf(ENV_VALUE_PREFIX) === 0) {
        const resolvedValueName: EnvironmentValue | undefined = NORMALIZED_ENV_VALUE_NAMES[normalizedEnvVar];
        if (resolvedValueName) {
          EnvironmentConfiguration._rushValues.set(resolvedValueName, process.env[envVar]);
        } else {
          unknownEnvVariables.push(envVar);
        }
      }
    }

    if (unknownEnvVariables.length > 0) {
      throw new Error(`Unknown environment variables: ${unknownEnvVariables.join(', ')}`);
    }
  }

  public static getEnvironmentValue(envValue: EnvironmentValue): string | undefined {
    return EnvironmentConfiguration._rushValues.get(envValue);
  }
}
