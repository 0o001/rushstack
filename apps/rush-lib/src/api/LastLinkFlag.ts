// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';

import { FlagFileBase } from './FlagFileBase';

export const LAST_LINK_FLAG_FILE_NAME: string = 'last-link.flag';

/**
 * The interface for the LastLinkFlag JSON file.
 *
 * @internal
 */
export interface ILastLinkFlagJson {}

/**
 * A helper class for managing the last-link flag, which is persistent and
 * indicates that linking was completed successfully.
 * @internal
 */
export class LastLinkFlag extends FlagFileBase<ILastLinkFlagJson> {
  /**
   * Creates a new LastLink flag
   * @param folderPath - the folder that this flag is managing
   */
  public constructor(folderPath: string, state: ILastLinkFlagJson = {}) {
    super(path.join(folderPath, LAST_LINK_FLAG_FILE_NAME), state);
  }

  /**
   * Returns true if the file exists and the contents match the current state.
   */
  public isValid(): boolean {
    const oldState: ILastLinkFlagJson | undefined = this.loadFromFile();
    return !!oldState;
  }
}
