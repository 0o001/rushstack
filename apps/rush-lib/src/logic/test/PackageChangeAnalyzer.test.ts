// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { PackageChangeAnalyzer } from '../PackageChangeAnalyzer';
import { RushConfiguration } from '../../api/RushConfiguration';
import { EnvironmentConfiguration } from '../../api/EnvironmentConfiguration';
import { RushConfigurationProject } from '../../api/RushConfigurationProject';

describe('PackageChangeAnalyzer', () => {
  beforeEach(() => {
    jest.spyOn(EnvironmentConfiguration, 'gitBinaryPath', 'get').mockReturnValue(undefined);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  function createTestSubject(
    projects: RushConfigurationProject[],
    files: Map<string, string>
  ): PackageChangeAnalyzer {
    const rushConfiguration: RushConfiguration = {
      commonRushConfigFolder: '',
      projects,
      rushJsonFolder: '',
      getCommittedShrinkwrapFilename(): string {
        return 'common/config/rush/pnpm-lock.yaml';
      },
      findProjectForPosixRelativePath(path: string): object | undefined {
        return projects.find((project) => path.startsWith(project.projectRelativeFolder));
      }
    } as RushConfiguration;

    const subject: PackageChangeAnalyzer = new PackageChangeAnalyzer(rushConfiguration);

    subject['_getRepoDeps'] = jest.fn(() => {
      return files;
    });

    return subject;
  }

  describe('getPackageDeps', () => {
    it('returns the files for the specified project', () => {
      const projects: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject,
        { packageName: 'banana', projectRelativeFolder: 'apps/banana' } as RushConfigurationProject
      ];
      const files: Map<string, string> = new Map([
        ['apps/apple/core.js', 'a101'],
        ['apps/banana/peel.js', 'b201']
      ]);
      const subject: PackageChangeAnalyzer = createTestSubject(projects, files);

      expect(subject.getPackageDeps('apple')).toEqual(new Map([['apps/apple/core.js', 'a101']]));
      expect(subject.getPackageDeps('banana')).toEqual(new Map([['apps/banana/peel.js', 'b201']]));
    });

    it('includes the committed shrinkwrap file as a dep for all projects', () => {
      const projects: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject,
        { packageName: 'banana', projectRelativeFolder: 'apps/banana' } as RushConfigurationProject
      ];
      const files: Map<string, string> = new Map([
        ['apps/apple/core.js', 'a101'],
        ['apps/banana/peel.js', 'b201'],
        ['common/config/rush/pnpm-lock.yaml', 'ffff'],
        ['tools/random-file.js', 'e00e']
      ]);
      const subject: PackageChangeAnalyzer = createTestSubject(projects, files);

      expect(subject.getPackageDeps('apple')).toEqual(
        new Map([
          ['apps/apple/core.js', 'a101'],
          ['common/config/rush/pnpm-lock.yaml', 'ffff']
        ])
      );
      expect(subject.getPackageDeps('banana')).toEqual(
        new Map([
          ['apps/banana/peel.js', 'b201'],
          ['common/config/rush/pnpm-lock.yaml', 'ffff']
        ])
      );
    });

    it('returns undefined if the specified project does not exist', () => {
      const projects: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject
      ];
      const files: Map<string, string> = new Map([['apps/apple/core.js', 'a101']]);
      const subject: PackageChangeAnalyzer = createTestSubject(projects, files);

      expect(subject.getPackageDeps('carrot')).toBeUndefined();
    });

    it('lazy-loads project data and caches it for future calls', () => {
      const projects: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject
      ];
      const files: Map<string, string> = new Map([['apps/apple/core.js', 'a101']]);
      const subject: PackageChangeAnalyzer = createTestSubject(projects, files);

      // Because other unit tests rely on the fact that a freshly instantiated
      // PackageChangeAnalyzer is inert until someone actually requests project data,
      // this test makes that expectation explicit.

      expect(subject['_data']).toBeNull();
      expect(subject.getPackageDeps('apple')).toEqual(new Map([['apps/apple/core.js', 'a101']]));
      expect(subject['_data']).toBeDefined();
      expect(subject.getPackageDeps('apple')).toEqual(new Map([['apps/apple/core.js', 'a101']]));
      expect(subject['_getRepoDeps']).toHaveBeenCalledTimes(1);
    });
  });

  describe('getProjectStateHash', () => {
    it('returns a fixed hash snapshot for a set of project deps', () => {
      const projects: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject
      ];
      const files: Map<string, string> = new Map([
        ['apps/apple/core.js', 'a101'],
        ['apps/apple/juice.js', 'e333'],
        ['apps/apple/slices.js', 'a102']
      ]);
      const subject: PackageChangeAnalyzer = createTestSubject(projects, files);

      expect(subject.getProjectStateHash('apple')).toMatchInlineSnapshot(
        `"265536e325cdfac3fa806a51873d927a712fc6c9"`
      );
    });

    it('returns the same hash regardless of dep order', () => {
      const projectsA: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject
      ];
      const filesA: Map<string, string> = new Map([
        ['apps/apple/core.js', 'a101'],
        ['apps/apple/juice.js', 'e333'],
        ['apps/apple/slices.js', 'a102']
      ]);
      const subjectA: PackageChangeAnalyzer = createTestSubject(projectsA, filesA);

      const projectsB: RushConfigurationProject[] = [
        { packageName: 'apple', projectRelativeFolder: 'apps/apple' } as RushConfigurationProject
      ];
      const filesB: Map<string, string> = new Map([
        ['apps/apple/slices.js', 'a102'],
        ['apps/apple/core.js', 'a101'],
        ['apps/apple/juice.js', 'e333']
      ]);
      const subjectB: PackageChangeAnalyzer = createTestSubject(projectsB, filesB);

      expect(subjectA.getProjectStateHash('apple')).toEqual(subjectB.getProjectStateHash('apple'));
    });
  });
});
