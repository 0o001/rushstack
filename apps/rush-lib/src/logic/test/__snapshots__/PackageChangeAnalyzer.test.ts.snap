// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageChangeAnalyzer getProjectStateHash returns a fixed hash snapshot for a set of project deps 1`] = `"265536e325cdfac3fa806a51873d927a712fc6c9"`;
